include/quickcpplib/algorithm/hash.hpp

```diff
@@ -48,12 +49,12 @@ namespace algorithm
     {
       size_t operator()(T v) const
       {
-#if defined(__x86_64__) || defined(_M_X64) || defined(__aarch64__) || defined(__ia64__) || defined(_M_IA64) || defined(__ppc64__)
-        static constexpr size_t basis = 14695981039346656037ULL, prime = 1099511628211ULL;
-        static_assert(sizeof(size_t) == 8, "size_t is not 64 bit");
-#else
+#if SIZE_MAX == 0xffffffff
```

Comment:
The problem here that you are in no way guaranteed that `size_t` has any relation to the platform's native bit size. It could be seventeen bits on a 64 bit platform, for example. In case you're thinking that would never happen, alas I've worked on a 64 bit platform where size_t had a 40 bit range. And that made sense, because only 40 bits were addressable on that platform.

`UINTPTR_MAX` is a much better choice, but some platforms don't define it. `__SIZEOF_POINTER__` is another good choice, but also not always present. Can I suggest you ask if either of those two is defined, if they are, use them, otherwise fall back to the previous logic, except adding in MSVC's macro for ARM64?

May I also suggest moving said macro logic into config.hpp, define something like `QUICKCPPLIB_PLATFORM_NATIVE_BITLENGTH` or your name of choice, and then we concentrate all the special logic in one place. Also, make sure the user can predefine your choice of macro naming, and then the default setting logic gets skipped.

Reply:
Apologies for the late reply. So little free time recently. I think that the hash function works in `size_t` because `std::hash` returns a `size_t`, and that is incorrect: it should work in a known sized integer type and cast to `size_t` only on return.

I guess the preprocessor logic could be put into config.hpp and made more intelligent and probe various options before falling back onto raw architecture matching. I'll try to find some time to get this done later this week.